great_ape_safe/ape_api/ebtc.py

```diff
@@ -866,3 +872,211 @@ def authority_set_authority(self, new_authority):
+        # ref: https://github.com/ebtc-protocol/ebtc/blob/main/packages/contracts/contracts/Interfaces/ICdpManagerData.sol#L91
+        cdp_id_status = self.cdp_manager.getCdpStatus(cdp_id)
+        assert cdp_id_status == CdpStatus.CLOSED.value
+
+        # 2.2. verify expected values are 0 at readings from the cdp manager
+        (
+            cdp_id_debt,
+            cdp_id_coll,
+            cdp_id_stake,
+            cdp_id_liq_reward_shares,
+            _,
+            _,
+        ) = self.cdp_manager.Cdps(cdp_id)
+        assert self.cdp_manager.cdpStEthFeePerUnitIndex(cdp_id) == 0
+        assert cdp_id_debt == 0
+        assert cdp_id_coll == 0
+        assert cdp_id_stake == 0
+        assert cdp_id_liq_reward_shares == 0
+
+    def cdp_add_collateral(self, cdp_id, coll_amount):
+        """
+        @dev Adds the received stETH to the specified Cdp.
+        @param cdp_id The CdpId on which this operation is operated.
+        @param coll_amount The total stETH collateral amount deposited (added) for the specified Cdp.
+        """
+        # verify: is it coll balance available
+        self._assert_collateral_balance(coll_amount)
+
+        # verify: cdp id ownership from caller
+        self._assert_cdp_id_ownership(cdp_id)
+
+        feed_price = self.price_feed.getPrice()
+        prev_icr = self.cdp_manager.getCachedICR(cdp_id, feed_price)
+        prev_tcr = self.cdp_manager.getCachedTCR(feed_price)
+        prev_coll_balance = self.cdp_manager.getCdpCollShares(cdp_id)
+
+        # 1. collateral approval for BO
+        self.collateral.approve(self.borrower_operations.address, coll_amount)
+
+        # 2. increase collateral in target cdp id
+        self.borrower_operations.addColl(cdp_id, cdp_id, cdp_id, coll_amount)
+
+        # 3. assertions:
+
+        # 3.1. icr and tcr had increased
+        assert self.cdp_manager.getCachedICR(cdp_id, feed_price) > prev_icr
+        assert self.cdp_manager.getCachedTCR(feed_price) > prev_tcr
+
+        # 3.2 collateral in cdp at storage has increased
+        assert self.cdp_manager.getCdpCollShares(cdp_id) > prev_coll_balance
+
+    def cdp_withdraw_collateral(self, cdp_id, coll_amount):
+        """
+        @dev Withdraws the amount of collateral from the specified Cdp.
+        @param cdp_id The CdpId on which this operation is operated.
+        @param coll_amount The total stETH collateral amount withdrawn (reduced) for the specified Cdp.
+        """
+        # verify: cdp id ownership from caller
+        self._assert_cdp_id_ownership(cdp_id)
+
+        feed_price = self.price_feed.getPrice()
+        prev_icr = self.cdp_manager.getCachedICR(cdp_id, feed_price)
+        prev_tcr = self.cdp_manager.getCachedTCR(feed_price)
+        prev_coll_balance = self.cdp_manager.getCdpCollShares(cdp_id)
+
+        # 1. decreased collateral in target cdp id
+        self.borrower_operations.withdrawColl(cdp_id, coll_amount, cdp_id, cdp_id)
```

Comment:
Would confirm that position holds enough collateral for the request before transacting to catch before reverting.

Reply:
sure, assertion makes sense! 

included in 432408bf3539e82fd68aa19b662963930dfc7828